contrib/hooks/lib.py

```diff
@@ -0,0 +1,139 @@
+"""
+Utility functions for hook scripts.
+"""
+
+import sys
+import os
+import subprocess
+import socket
+
+from subprocess import CalledProcessError, TimeoutExpired
+from typing import Tuple, Optional
+
+def complain(message):
+    sys.stderr.flush()
+    sys.stderr.write(message)
+    sys.stderr.write('\n')
+    sys.stderr.flush()
+
+def announce(message):
+    sys.stderr.flush()
+    sys.stderr.write(message)
+    sys.stderr.write('\n')
+    sys.stderr.flush()
+
+cached_hostname = None
+def get_hostname() -> str:
+    global cached_hostname
```

Comment:
Suggestion: use `functools.lru_cache()` here instead of a global, and a warning log message if it fails.  i.e.

```
@functools.lru_cache
def get_hostname() -> str:
    try:
        return socket.getfqdn()
    except Exception as e:
        warning.warn(f'An error occurred while attempting to describe the host: {e}')
    return 'localhost'

Reply:
I didn't want to use this because it's fiddly between versions; what you wrote won't work on 3.7, for example, because the max size (or maybe just some parens) are required.

But I can swap over.